### PR TITLE
test: sharpen unit and integration suites

### PR DESCRIPTION
**Pull request by rmens**
Full review of both test layers (Go unit tests and the CI integration suites) with one goal: every test either proves something real, or it goes. Net −612 lines while coverage goes up.

## Unit tests

**Speed: `go test ./...` drops from ~9.7s to ~2.3s**
- New `statTimeoutFor` seam in `filemonitor.go` (house style, next to `osStat`/`nowFunc`): six tests each burned a real 1.00s on the smallest configurable stat timeout. `internal/service`: 7.5s → 0.9s.
- `internal/async` sleeps replaced with channel synchronization: 0.9s → 0.1s. The rewritten checks are also stronger (they now *prove* `Close()` blocked instead of racing a 50ms sleep).

**Deleted tests that could not fail or duplicated a stronger sibling**
- A test of `encoding/json` against a locally declared struct, a test of `sync.WaitGroup` semantics, a torn-write test that only read quiescent state, and four subset-scenarios fully contained in a neighbouring test.

**Sharper assertions**
- Typed errors (`errors.AsType[*types.ValidationError]` + `Field`) instead of error-message substrings.
- Config validation tests now assert the field label (`filemonitor.checks[0].path`, …) so they can't pass on a neighbouring rule firing.
- Rate-limiter and API-key tests assert behavior through the public API/router instead of peeking at private maps; the permission-denied test uses the `osStat` seam instead of `chmod 0000` (was failing under root).
- Removed constant-vs-constant assertions (`Getter() == DefaultConstant` where the getter is `cmp.Or(x, DefaultConstant)`).
- `buildTestMatcher` now mirrors production `buildMatcher` wiring, so tests exercise real configurations.

**Dedup**: one backup-service constructor (was two with transposed names), one server bootstrap helper (was 11 copies), one hanging-stat stub (was two), shared channel-wait/file-write helpers, and table-driven consolidation across api/config/database/service/async/notify. `internal/api` tests now run with `t.Parallel()`.

**One small production addition**: `async.Runner.Closing()`, so tests can observe shutdown without spawning probe goroutines that mutate what they measure.

## Integration suites (comprehensive-test.yml)

**Real bug fixed**: the fixture seeded artist/track images as `E'\\\\x89504e...'`, which stores the ASCII *text* `\x89504e...` (escape-format bytea), not PNG bytes — and no test noticed because images were only fetched after a fresh upload. Now a plain hex literal (verified against a live postgres:17 container: `octet_length=8`), seeded deterministically (exactly 50+50), with a new regression step that GETs a fixture-seeded image and checks the PNG magic.

**Checks that verified less than they appeared to**
- "Long-running query detection" gated on `count >= 0` (unfalsifiable) — deleted; its structure assertions moved to the low-threshold step where detection is guaranteed.
- Scheduler step only emitted `::warning::` — now fails on missing log lines and asserts a DB-backed endpoint instead of re-checking `/health`.
- `database` suite tested the fixture, not the product — folded into the shared DB gate (now asserting exact counts); suite removed from the matrix.
- Concurrent-backup 409 test polled for a state the API guarantees synchronously and hard-failed when the backup finished fast — now POST → 202, POST → 409, deterministic.

**CI cost/flake**
- MinIO behind a compose `profile`, started only for the s3 suite (was booted in all 9 jobs, then destroyed and hand-rebuilt by the one job using it).
- Fixed sleeps → polls; silent readiness loops now fail their own step with a server-log dump; `timeout-minutes` added; imagemagick install gated to the 3 suites using it.
- Deleted: no-op cleanup steps, unused `test_small.png`, unreferenced `tests/Dockerfile.testdb`, duplicate endpoint checks; README fixture counts corrected (1000/1100, not 80/130).

**New one-line sibling checks**: authenticated `GET /api/health` (incl. `-ldflags` version wiring), tracks bulk-delete confirmation + image-404-after-delete (mirroring artists), `POST /api/notifications/test-email` → 400, disabled-backup routes → 404, `HEAD` support.

## Not done here (follow-up candidates)
- Extracting the 9 suite bodies from YAML into `tests/suites/*.sh` with a shared assert library (~400 more lines of duplication; would also make suites runnable locally and shellcheck-able).
- Trimming dead fixture weight (628 orphaned playlistitem rows and unused fadelist blobs, ~24% of mock_data.sql).

Verified: `go build`, `go vet`, `gofmt`, `golangci-lint` (0 issues), `staticcheck`, `deadcode`, full `go test ./... -count=1` and `-race`; workflow YAML parses and actionlint reports no new findings vs the pre-edit baseline.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved reliability for shutdown handling, timeouts, concurrent operations, backup cleanup, file monitoring, media matching, and image validation.
  * Strengthened API response safety, authentication, rate-limit behavior, and notification checks.

* **Tests**
  * Expanded automated coverage for APIs, configuration, storage, scheduling, maintenance, backups, and uploads.
  * Added deterministic fixtures, asynchronous operation checks, and stricter validation.

* **Documentation**
  * Updated local testing instructions with optional object-storage support and consistent port configuration.

### CODE REVIEW

File: .github/workflows/comprehensive-test.yml [hunk 1]
```diff
@@ -87,25 +88,26 @@ jobs:
...
+          if [ "${{ matrix.suite }}" = "artists" ] || [ "${{ matrix.suite }}" = "tracks" ] || [ "${{ matrix.suite }}" = "edge-cases" ]; then
+            sudo apt-get install -y imagemagick
           fi
 
       - name: Start test database
         run: |
           cd tests
-          docker compose -f docker-compose.test.yml up -d --wait
+          # MinIO sits behind the "s3" compose profile; only the s3 suite needs it.
+          docker compose -f docker-compose.test.yml ${{ matrix.suite == 's3' && '--profile s3' || '' }} up -d --wait
```
[the remark this answers is not in the document]
Comment (rmens) [reply .github/workflows/comprehensive-test.yml]: Gedeeltelijk eens met het algemene risico: `latest` is mutable. Deze tag bestond echter al op `main`; deze PR voegt alleen het bestaande MinIO-serviceprofiel toe aan de S3-suite. Voor deze testopschonings-PR hoeft een nieuwe pin plus bijbehorend updatebeleid niet te bestaan (YAGNI/scope). Ik laat dit daarom hier staan en zou image-pinning als aparte dependency-maintenancewijziging doen.

File: .github/workflows/comprehensive-test.yml [hunk 2]
```diff
@@ -994,43 +1011,29 @@ jobs:
...
+          HTTP_CODE=$(curl -s -o /tmp/concurrent_backup.json -w "%{http_code}" -X POST \
+            -H "X-API-Key: ${{ env.API_KEY }}" \
+            -H 'Content-Type: application/json' \
+            http://localhost:${{ env.API_PORT }}/api/db/backup \
+            -d '{}')
+          if [ "$HTTP_CODE" = "409" ] && jq -e '.success == false' /tmp/concurrent_backup.json > /dev/null; then
+            echo "Correct: concurrent backup afgewezen (HTTP 409)"
+          else
+            echo "::error::Verwachtte HTTP 409 met success=false, kreeg HTTP $HTTP_CODE"
+            cat /tmp/concurrent_backup.json
```
[the remark this answers is not in the document]
Comment (rmens) [reply .github/workflows/comprehensive-test.yml]: Eens met de diagnose, niet met extra synchronisatie in deze integratiestap. Het slot is bij de eerste response gereserveerd, maar kan vóór de tweede request alweer vrij zijn; de 409-check blijft dus timing-afhankelijk. De minimale deterministische dekking hergebruikt de bestaande lagen: reserveer `svc.runner` in een `BackupService.Start`-unittest en assert de `ConflictError`; `internal/api/response_test.go` dekt al `ConflictError` → HTTP 409. Dan kan deze timingtest weg zonder een test-only blokkade aan de API toe te voegen.

File: .github/workflows/comprehensive-test.yml [hunk 3]
```diff
@@ -1625,30 +1613,10 @@ jobs:
...
-          jq -e '.success == true' /tmp/r3.json > /dev/null || FAILED=1
-
-          if [ "$FAILED" = "0" ]; then
+          if jq -es 'map(.success) | all' /tmp/r1.json /tmp/r2.json /tmp/r3.json > /dev/null; then
             echo "Alle 3 gelijktijdige uploads succesvol"
           else
             echo "::error::Een of meer gelijktijdige uploads gefaald"
+            jq -s '.' /tmp/r1.json /tmp/r2.json /tmp/r3.json
             exit 1
           fi
```
[the remark this answers is not in the document]
Comment (rmens) [reply .github/workflows/comprehensive-test.yml]: Eens. Een leeg curl-resultaat levert bij `jq -s` geen element op en `all` is voor een lege of te korte verzameling vacuüm waar. `length == 3 and (map(.success) | all)` is hier precies de minimale, éénregelige aanscherping.

File: internal/config/mediacheck_test.go [hunk 4]
```diff
@@ -1,110 +1,112 @@
...
+			}
+			if err == nil {
+				t.Fatalf("validate() error = nil, want message containing %q", tt.wantErrContains)
+			}
+			if !strings.Contains(err.Error(), tt.wantErrContains) {
+				t.Errorf("validate() error %q missing %q", err.Error(), tt.wantErrContains)
+			}
+		})
 	}
 }
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/config/mediacheck_test.go:100]: Niet mee eens. Dit zijn twee lokale table-testloops met elk eigen arrange-stap; een packagebrede helper introduceert nu een derde concept en cross-file koppeling zonder extra gedrag of derde gebruiker. De assertions naast de cases houden is hier leesbaarder en volgt YAGNI; bij een derde echte gebruiker kunnen we alsnog extraheren.

File: internal/database/mediacheck_test.go [hunk 5]
```diff
@@ -1,125 +1,132 @@
...
+	"as trackid",
+	"as tracktitle",
+	"as artist",
+	"as start_time",
+	"as blockid",
+	"as block",
+	"as filepath",
+	"as filename",
+	"as audioname",
 }
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/database/mediacheck_test.go:24]: Eens met de robuustheid, al is dit geen functionele fout: `mediaCheckColumns` gebruikt nu aantoonbaar lowercase `as`. SQL-keywordcasing is niet semantisch, dus één keer `query` met stdlib `strings.ToLower` normaliseren voorkomt dat een pure formatteringswijziging deze aliascontracttest breekt. Dat is de minimale oplossing.

File: internal/database/mediacheck_test.go [hunk 6]
```diff
@@ -1,125 +1,132 @@
...
+			opts:            MediaCheckOptions{IncludeVoicetracks: true},
+			wantNotContains: []string{"t.userid <>"},
+			wantParams:      nil,
+		},
+		{
+			name:         "limit is bound last",
+			opts:         MediaCheckOptions{Limit: 50},
+			wantContains: []string{"LIMIT"},
+			wantParams:   []any{types.VoicetrackUserID, 50},
+		},
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/database/mediacheck_test.go:91]: Eens. De bestaande parametervergelijking bewijst de waardenvolgorde, maar niet dat `Limit` daadwerkelijk via de tweede placeholder in de SQL terechtkomt. `LIMIT $2` maakt de case-naam waar en is een éénregelige aanscherping.

File: internal/database/playlist_test.go [hunk 7]
```diff
@@ -14,116 +14,86 @@ func TestBuildPlaylistQueryRequiresBlockID(t *testing.T) {
...
+			},
+			wantContains: []string{
+				"AND t.picture IS NOT NULL",
+				"AND a.picture IS NULL",
+				"LIMIT $2",
+			},
+			wantNotContains: []string{"AND t.picture IS NULL", "AND a.picture IS NOT NULL", " OFFSET "},
+			wantParams:      []any{"block-1", 25},
 		},
 	}
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/database/playlist_test.go:76]: Niet mee eens; deze dekking bestaat al in de huidige branch. De tabelcase op regels 41–58 assert `ORDER BY t.artist DESC`, `TestBuildPlaylistQuerySortByDuration` dekt duration, en `TestBuildPlaylistQuerySortByIsAllowListed` voert juist een onverwachte/malafide `SortBy` in en assert de veilige fallback `ORDER BY pi.startdatetime DESC`. `internal/service/media_test.go` behoudt daarnaast de default `start_time`. Er is dus niets verloren en geen extra test nodig.

File: internal/image/image_test.go [hunk 8]
```diff
@@ -13,23 +13,32 @@ import (
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := Process(tt.data, Config{TargetWidth: 10, TargetHeight: 10, Quality: 85})
+			if err == nil {
+				t.Fatal("Process accepted invalid image data")
+			}
+			if !strings.Contains(err.Error(), "failed to get image information") {
+				t.Fatalf("error = %q, want image information failure", err)
+			}
+		})
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/image/image_test.go:40]: Niet mee eens met de voorgestelde abstractie. Beide inputs falen momenteel bewust in stdlib `image.DecodeConfig` en krijgen daardoor exact hetzelfde productieprefix `failed to get image information`; een `wantErrContains`-veld met tweemaal dezelfde waarde voegt geen signaal toe. De verschillende input plus de non-nil error bewijst al dat GIF niet wordt geaccepteerd. Een GIF-specifieke melding zou eerst nieuw productiegedrag vereisen en valt buiten deze testrefactor.

File: internal/service/mediamatch_test.go [hunk 9]
```diff
@@ -94,247 +115,243 @@ func buildTestMatcher(t *testing.T, driveDirs map[string]string, searchDirs []st
...
-	if out.Status != MediaStatusMissing {
-		t.Fatalf("status = %q, want missing", out.Status)
-	}
-	if len(out.Matches) != 0 {
-		t.Fatalf("matches = %v, want none", out.Matches)
+		<-releaseCh
+		return info, err
 	}
+	return startedCh, releaseFn, counter
 }
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/service/mediamatch_test.go:167]: Niet mee eens na het traceren van de echte flow. `statRootWithTimeout` kopieert eerst `mediaRootStat` naar lokaal `statFn`, en `buildIndex` doet hetzelfde met `mediaWalkDir`; de achterblijvende goroutines voeren die gekopieerde functie uit en lezen de packagevariabele niet opnieuw. De match-caller is vóór cleanup al terug, terwijl omgekeerde cleanupvolgorde eerst de stub vrijgeeft en daarna herstelt. De gerichte timeout/cancel-tests zijn bovendien 25× schoon onder `go test -race`. Een extra `WaitGroup` is hier dus onnodig en maakt de stub complexer.

File: internal/service/mediamatch_test.go [hunk 10]
```diff
@@ -356,49 +373,30 @@ func TestMatch_StatTimeoutUsesSingleFlight(t *testing.T) {
...
+	mustReceive(t, started, "first stat start")
 
 	go func() {
 		time.Sleep(10 * time.Millisecond)
-		closeRelease()
+		release()
 	}()
 	second := m.match(input)
-	firstOut := receiveMatchOutcome(t, first)
+	firstOut := mustReceive(t, first, "first match outcome")
```
[the remark this answers is not in the document]
Comment (rmens) [reply internal/service/mediamatch_test.go:411]: Eens dat de `10ms`-sleep geen geldige join-garantie is; oneens dat hiervoor productiecode of een waiter counter nodig is. De test heeft al de juiste seam: `m.startStatFlight`. Wikkel `flights.startOrJoin` test-lokaal, signaleer wanneer de tweede call `isNew == false` terugkrijgt, laat de tweede `match` in een goroutine lopen, wacht op dat signaal en roep dan `release()` aan. Daarmee wordt de test deterministisch zonder nieuw productieoppervlak.

**Comment by rmens** [comment]
@coderabbitai full review

**Comment by rmens** [comment]
@coderabbitai Follow-up na push van commit `33490a2`:

Opgevolgd:
- De API-keycheck vereist nu expliciet `.success == true`.
- De twee Runner-shutdowntests wachten nu deterministisch op het bestaande shutdownsignaal. De reservatietest reserveert vóór `Close`, start `Go` pas nadat shutdown begonnen is en bewijst dat `Close` op het gereserveerde werk wacht.
- De fixture bevat nu een complete, decodeerbare 1x1-PNG. De workflow laat de reeds geïnstalleerde ImageMagick-decoder het opgehaalde bestand als PNG valideren in plaats van alleen magic bytes te vergelijken.
- De outside-diff-suggestie is verwerkt in `TestCheckResult_StateProfiles`: de echte `FileCheckResult` wordt gemarshald en `file_exists` wordt voor `true`, `false` en `null` gecontroleerd. Dit staat bewust in package `service`, waar het productie-API-type leeft, niet als nagebouwd structje in `config_test.go`.

Niet overgenomen:
- De epsilon voor `DeadTuplePct` is niet nodig. Productie en test voeren voor deze ongeronde contractwaarde exact dezelfde `float64`-operaties in dezelfde volgorde uit; de vergelijking is deterministisch. De voorgestelde `1e-9` zou het genoemde afronden op twee decimalen bovendien nog steeds terecht laten falen. Een extra `math`-import voegt hier dus geen bescherming toe.
- `minio/minio:latest` blijft ongewijzigd. Die referentie staat al op `main`; deze PR voegt alleen het bestaande S3-profiel toe en introduceert de mutable tag niet. Een versiepin is een afzonderlijke dependency-keuze en valt buiten deze testverscherping.

Verificatie: `go test -race ./...`, `go vet ./...`, `docker compose config`, workflow-YAML-parse en een echte ImageMagick-decode van de PNG uit de SQL-literal zijn allemaal geslaagd.

File: internal/config/config_test.go [hunk 11]
```diff
@@ -4,167 +4,194 @@ import (
...
 		{name: "enabled with minimum-length key", enabled: true, keys: []string{"0123456789abcdef"}},
-		{name: "enabled with nil keys", enabled: true, wantErr: true},
-		{name: "enabled with empty keys", enabled: true, keys: []string{}, wantErr: true},
-		{name: "enabled with empty key", enabled: true, keys: []string{""}, wantErr: true},
-		{name: "enabled with short key", enabled: true, keys: []string{"short-key"}, wantErr: true},
-		{name: "disabled with short key", enabled: false, keys: []string{"short-key"}, wantErr: true},
+		{
+			name:            "enabled with nil keys",
+			enabled:         true,
+			wantErrContains: "api.keys must have at least one entry when enabled",
```
Comment (rmens) [inline internal/config/config_test.go:169]: @coderabbitai Agreed and addressed in 8703bbf. Each negative case now asserts the specific API-key validation message (missing keys, empty element, or short element), so an unrelated failure in minimalConfig can no longer make the test pass. I kept the existing string-based contract: validate() intentionally converts validator errors to user-facing text through formatErrors(), and the surrounding validation tests use the same pattern. Verified with the focused test (10 runs), the full race-enabled suite, and go vet.
